primer.md

```diff
@@ -112,11 +112,36 @@ specifications of how to serialize the event in different formats and
 transports (e.g. JSON and HTTP).
 
 ### Non-Goals
-The following will not be part of the specification:
+The following are considered beyond the scope of the specification:
 * Function build and invocation process
 * Language-specific runtime APIs
 * Selecting a single identity/access control system
 
+The following are features that the working group has explicitly decided
+should nto be part of the specification:
+* **routing information** The CloudEvents spec will never include routing
```

Comment:
Since this covers routing, should it be added that data relevant to routing can( should?) be put in `extensions`( or whatever it ends up being named)? Things like partition key or @cathyhongzhang IoT usecase are routing-relevant and the current phrasing seems to completely deny the `extensions` option.

Reply:
That's good point. Just because this WG decided that we didn't want the core set of attributes to include routing doesn't mean that people couldn't add them in via extensions. We may, or may not, think its a good idea but people are free to do just about anything they want with extensions.  It is fair for us to provide guidance on why such things might be bad though - which I think he already does.

Reply:
That almost begs the question... does it belong in `data` then?

I'm actually not suggesting that the spec should say where it should go either way because I think people should be able to add any extensions they want to our spec, but for use cases where they want to do processing out-of-bounds of the transferring of messages between two components, it feels less like a CloudEvent problem to me. But this goes back to what kind of guidance we provide for people on when to use CE extensions vs use `data`.

Reply:
You are confusing me:-)  CloudEvents are not just dealing with transporting/transferring of messages between two components. I think one of its goals is to facilitate interpretation and  processing of the event on the event consuming side. Adding this identificationLabels attribute to the main spec.md will help the event consuming side to handle a large number of serverless use cases/applications that involve multiple events, such as those IoT use cases, otherwise the event consumer might need to decrypt and decode the needed info from a big payload (I remember someone raised such a concern before).  Also the needed info might not be in the payload.

BTW, I would suggest that we add a discussion topic in next meeting on whether we allow different classifications/categorizations/bags of event information.

Reply:
> Since this covers routing, should it be added that data relevant to routing can( should?) be put in extensions( or whatever it ends up being named)? Things like partition key

I maintain that routing information is _never_ a good idea, even in extensions. This was actually one of the reasons why I shot down "partitionKey" or "correlationId". The event should have enough information that the _application_ can choose what to partition or correlate by.

Now that I can be more concrete, if one looks at KNative eventing, it's totally reasonable that a the Flow/Feed adds a feature for correlation or partitioning. Then, the actual _Flow_ could set a correlation strategy of "event.labels.house" for one application and "event.labels.complex" for another. If the event itself is making the decisions for how it should be routed, these use cases are broken.

Reply:
Just want to add on Thomas's point. For the correlation usage, the event itself does not know which identification label carried in the event metadata will be used to correlate it with other events. It is up to another entity (could be the use case application developer) to set the correlation strategy as, e.g., correlationID=event.labels.house or correlationID=event.identificationLabels.house assuming we name it "identifcationLabels" instead of "Labels".

Reply:
Doug: The event producer does not know how the event consumer will use the identification label or for what usage, just like it does not know how the event consumer will use the eventType.  But the event producer knows what Labels can be used to identify this event.

Reply:
"can" is the key word. It also "can" be used for correlation, logging, debugging, etc... just about anything.  The producer has no idea how any of these will be used. So trying to put them into classifications doesn't seem very useful since it would just be a guess anyway. As you implied... eventType is in the same position - so why don't we put it into their too?  

Actually, I'm suggesting we do ;-)   Just the other way around.  Put these things into the same "can be used for anything" category as all of the spec-defined CloudEvent properties.

The meaning of the property is defined by its spec - that's all the consumer will care about, not whether we wrap it in some bag, along with other properties they may or may not care about, or that may or may not be used for the same purpose.

Reply:
@duglin 
>It also "can" be used for correlation, logging, debugging, etc... just about anything.

Here's my sargument for having a bucket/bag/container for additional identification.
Consider the three cases below.  Case1 does not have an identificationLabel while Case2 and Case3 do have identificationLabel.

Functionality wise all three cases are the same for the developer writing the app.  The correlation 
for Case1 is event.taste whilst for Case2/Case3 the correlation is for event.identificationLabel.taste

Now say you are the customer using the app and you can see the events but don't know anythng beyond that.  If you do not know what the key event identifier is where do you start?  Do you classify based on 'color' or 'house'.  But 'house' is not even present in Case3.   
On the other hand, Case2 and Case3 clearly illustrate that the key event identifier is 'taste'. You then classify the events based on 'taste'.  
taste: bitter, total events 131450
taste: sour,  total events 451345
taste: sweet, total events 465000

Case1:
eventID: 0x7145
color: red
house: small
taste:  bitter
boat: big
car: used

Case2:
eventID: 0x7145
animal: horse
house: small
identificationLabel: 
  taste: sour
boat: small
car: new

Case3:
eventID: 0x7145
color: red
animal: dog
owner: yes
record: 1234
identificationLabel: 
  taste: sweet

Reply:
@fsunaval Any place people can add "stuff", whether its at the top level or in a bag, means you need to be prepared for "stuff" you don't care about. This means you're assumption that someone will do some kind of semantic processing (e.g. classifying) based on a bag of "random" stuff seems pretty unlikely to me. There could be 100's of properties in that bag that we don't care about, we just don't know, its open ended.  Even @cathyhongzhang agreed that the producer has no idea which properties the consumers will care about, this means they'll be putting all sorts of stuff in there because different consumers might use different properties and they will probably need to include everything that they have just in case someone needs it. 

In order to do something meaningful with the data the consumer must know what to look for - to find the gems in that random bag of stuff. Once that assumption is agreed to, where those properties are isn't really relevant. They can just as easily find them at the top-level as they can in a bag.

Now, let's say someone does really want to have a consumer be unaware of things - like you're suggesting. That's fine. But in that case I would claim that the producer and consumer would need to be in-sync in some way and have agreed on what "bag" to look for and that the only things in that bag will be exactly what the consumer needs and not some random other bits.  At that point I would not call that a "bag" as much as I would call it an extension property whose value is a structure with, most likely, a set of well-defined properties. Or at least some kind of constraints on it to make sure the only things in there are the things the "blind" consumer will need in its processing and not random "stuff" - e.g. a structure w/o an extensibility point. This is similar to what @inlined was asking about on today's call.

But the point is that we can not define this kind of structure in our spec today w/o being so focused on one particular usecase that it would either make this bag unusable for other cases, or lead us down the path of defining a lot of them based on every variant of usecase out there.  I would much prefer to let people define these specialized properties/bags as extensions based on their needs and **if** one day we realized that one of them became so popular that most people seem to be using it then we can promote it as a core spec property.  And because its already at the top-level no code would need to change - as a bonus  :-)

Reply:
Doug: I do not think that an event producer will put all sorts of stuff in the idenficationLabel bag. It is not a random bag of stuff. Of source we can not prevent anyone to do bad practice and bad practice can be done to any metadata field. When it comes down to each individual event producer which fills in the event metadata specific to that event type, the number of identification labels for that event will not be large. I don't feel your assumption is correct.

Reply:
All they have to do is put one item in the bag that some consumer doesn't care about and it makes my point.